### PR TITLE
Feature 9 render html as flutter widgets

### LINKED ISSUE #9: Render HTML as Flutter Widgets

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #9

### CODE REVIEW

File: lib/utils/html_widgets.dart [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  /// Builds custom Image widget from HTML [element] and applies [resources]
+  ///
+  /// TODO: Implement loading images from device memory
+  static Widget _buildCustomImageWidget(
+    dom.Element element,
+    List<ExhibitionPageResource> resources,
+    List<ExhibitionPageEventTrigger> eventTriggers,
+  ) {
+    Size size = _extractSize(element);
+    ExhibitionPageEventTrigger? foundEventTrigger =
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart]: There are many kinds of tiggers, should this be renamed as e.g. viewClickEventTrigger?

File: lib/utils/html_widgets.dart [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        _findEventTrigger(element, eventTriggers);
+    String? foundResource = resources
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart:42]: shouldn't it be @resources/${id}?

Comment (anttileppa) [reply lib/utils/html_widgets.dart:42]: and isn't the id stored as src -attribute?

File: lib/utils/html_widgets.dart [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  /// Event handler for tap events on custom widgets per [eventTrigger]
+  ///
+  /// TODO: Implement this
+  static Function() _handleTapEvent(ExhibitionPageEventTrigger eventTrigger) {
+    return () {};
+  }
+
+  /// Extracts elements width and height
+  static Size _extractSize(dom.Element element) {
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart:132]: This seems to handle size as pixels, is it possible that size is sometimes handled as e.g. percents?

File: lib/utils/html_widgets.dart [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  /// Builds custom Button widget from HTML [element] and applies [resources]
+  static Widget _buildCustomButtonWidget(
+    dom.Element element,
+    List<ExhibitionPageResource> resources,
+    List<ExhibitionPageEventTrigger> eventTriggers,
+  ) {
+    Size size = _extractSize(element);
+    ExhibitionPageEventTrigger? foundEventTrigger =
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart]: Maybe this should also be renamed?

File: lib/utils/html_widgets.dart [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        _findEventTrigger(element, eventTriggers);
+    String? foundResource = resources
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart:72]: shouldn't this also it be prefixed with @resources?

File: lib/utils/html_widgets.dart [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        .firstWhereOrNull((resource) => resource.id == element.innerHtml)
+        ?.data;
+
+    return ElevatedButton(
+        style: ButtonStyle(fixedSize: MaterialStateProperty.all(size)),
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart:79]: Styles should probably be derived from HTML element (can also be fixed with TODO-comment)

File: lib/utils/html_widgets.dart [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  /// Builds custom Video widget from HTML [element] and applies [resources]
+  static Widget? _buildCustomVideoWidget(
+    dom.Element element,
+    List<ExhibitionPageResource> resources,
+    List<ExhibitionPageEventTrigger> eventTriggers,
+  ) {
+    Size size = _extractSize(element);
+    ExhibitionPageEventTrigger? foundEventTrigger =
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart]: Maybe this should also be renamed?

File: lib/utils/html_widgets.dart [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        height: size.height,
+        child: VideoWidget(src: foundResource ?? ""),
+      );
+    }
+
+    return null;
+  }
+
+  /// Finds event trigger assigned to this [element] from [eventTriggers]
+  static ExhibitionPageEventTrigger? _findEventTrigger(
```
Comment (anttileppa) [inline lib/utils/html_widgets.dart]: This finds only clickView triggers, so it should be renamed to match the behaviour, e.g. findClickViewEventTrigger

**Review by anttileppa — APPROVED** [review]

File: lib/utils/html_widgets.dart (hunk 1 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart]: Resolved

File: lib/utils/html_widgets.dart (hunk 2 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart:42]: Resolved

File: lib/utils/html_widgets.dart (hunk 3 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart:132]: We had some thoughts of resolving all of percentage dimensions into pixels on management side. Lets discuss this in person.

File: lib/utils/html_widgets.dart (hunk 4 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart]: Resolved

File: lib/utils/html_widgets.dart (hunk 6 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart:79]: Resolved with TODO comment

File: lib/utils/html_widgets.dart (hunk 7 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart]: Resolved

File: lib/utils/html_widgets.dart (hunk 8 above)
Comment (villejuutila) [reply lib/utils/html_widgets.dart]: Resolved

**Pull request closed without merging**